File: leetcode/src/0100-0199/0146. LRU Cache.cs
using System;
using System.Collections.Generic;

namespace coding.leetcode {
    public class LRUCache {
        private readonly Dictionary<int, LinkedListNode<int>> addr;
        private readonly Dictionary<int, int> vals;
        private readonly LinkedList<int> list;
        private readonly int capacity;

        public LRUCache(int capacity) {
            this.capacity = capacity;
            this.addr = new Dictionary<int, LinkedListNode<int>>();
            this.vals = new Dictionary<int, int>();
            this.list = new LinkedList<int>();
        }

        public int Get(int key) {
            if (Warm(key)) {
                return vals[key];
            } else {
                return -1;
            }
        }

        public void Put(int key, int value) {
            if (Warm(key) == false) {
                addr[key] = list.AddFirst(key);
                if (list.Count > capacity) {
                    var last = list.Last;
                    addr.Remove(last.Value);
                    vals.Remove(last.Value);
                    list.Remove(last);
                }
            }
            vals[key] = value;
        }

        private bool Warm(int key) {
            if (addr.TryGetValue(key, out var node)) {
                list.Remove(node);
                list.AddFirst(node);
                return true;
            } else {
                return false;
            }
        }
    }
}

File: leetcode/src/0100-0199/0146. LRU Cache.java
public class LRUCache {
    private final LRUListNode<Integer, Integer>[] addr;
    private final LRUList<Integer, Integer> list;
    private final int capa;

    public LRUCache(int capacity) {
        this.capa = capacity;
        this.addr = new LRUListNode[10000 + 1];
        this.list = new LRUList<>();
    }

    public int get(int key) {
        var node = warm(key);
        if (node != null) {
            return node.val;
        }
        return -1;
    }

    public void put(int key, int val) {
        var node = warm(key);
        if (node == null) {
            node = new LRUListNode<>();
            list.insert(node);
            addr[key] = node;
            if (list.size > capa) {
                var last = list.last();
                addr[last.key] = null;
                list.remove(last);
            }
        }
        node.key = key;
        node.val = val;
    }

    private LRUListNode<Integer, Integer> warm(int key) {
        var node = addr[key];
        if (node != null) {
            list.remove(node);
            list.insert(node);
        }
        return node;
    }

    private final class LRUList<K, V> {
        public LRUListNode<K, V> head;
        public LRUListNode<K, V> tail;
        public int size;

        public LRUList() {
            this.head = new LRUListNode<>();
            this.tail = head;
            this.head.next = tail;
            this.tail.prev = head;
            this.size = 0;
        }

        public void insert(LRUListNode<K, V> node) {
            head.next.prev = node;
            node.next = head.next;
            node.prev = head;
            head.next = node;
            size = size + 1;
        }

        public void remove(LRUListNode<K, V> node) {
            var prev = node.prev;
            var next = node.next;
            prev.next = next;
            next.prev = prev;
            size = size - 1;
        }

        public LRUListNode<K, V> last() {
            return tail.prev;
        }
    }

    private final class LRUListNode<K, V> {
        public LRUListNode<K, V> prev;
        public LRUListNode<K, V> next;
        public K key;
        public V val;
    }
}